src/number_theory/primorial.lean

```diff
@@ -0,0 +1,195 @@
+/-
+Copyright (c) 2020 Patrick Stevens. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Stevens
+-/
+import tactic
+import data.nat.parity
+
+/-!
+# Primorial
+
+This file defines the primorial function, and proves a bound on its size.
+
+## Notations
+
+We use the notation `n#` for the primorial of `n`: that is, the product of the primes less than
+or equal to `n`.
+-/
+
+open finset
+open nat
+open_locale big_operators
+
+def primorial (n : ℕ) : ℕ := ∏ p in (filter prime (range (n + 1))), p
+local notation x`#` := primorial x
+
+-- This is going into stdlib
+lemma sum_range_choose_halfway (m : ℕ) : ∑ i in range (m + 1), choose (2 * m + 1) i = 4 ^ m :=
+   by sorry
+
+private lemma choose_symm_half (m : ℕ) : choose (2 * m + 1) (m + 1) = choose (2 * m + 1) m :=
+by apply choose_symm_of_eq_add; ring
+
+lemma primorial_succ {n : ℕ} (n_big : 2 < n) (r : n % 2 = 1) : (n + 1)# = n# :=
+begin
+  have not_prime : ¬prime (n + 1),
+    { intros is_prime,
+      cases (prime.eq_two_or_odd is_prime),
+      { linarith, },
+      { rw ←not_even_iff at h r,
+        have e : even (n + 1 - n), exact (even_sub (le_of_lt (lt_add_one n))).2 (iff_of_false h r),
+        exfalso,
+        simp only [nat.add_sub_cancel_left, not_even_one] at e,
+        exact e, }, },
+  apply finset.prod_congr,
+  { rw [@range_succ (n + 1), filter_insert, if_neg not_prime], },
+  { exact λ _ _, rfl, },
+end
+
+private lemma range_contains_upper_bound (n : ℕ) : n ∈ range (n + 1) :=
+by simp only [succ_pos', lt_add_iff_pos_right, mem_range]
+
+private lemma select_from_sum (f : ℕ → ℕ) (s : finset ℕ) (a : ℕ) (t : a ∈ s) : f a ≤ ∑ i in s, f i :=
+by rw <- finset.insert_erase t; simp
+
+private lemma bound_binomial_coefficient (n : ℕ) : choose (2 * n + 1) n ≤ 4 ^ n :=
+begin
+  have t : choose (2 * n + 1) n ≤ ∑ i in range (n + 1), choose (2 * n + 1) i,
+    exact select_from_sum (choose (2 * n + 1)) (range (n + 1)) n (range_contains_upper_bound n),
+  simpa only [sum_range_choose_halfway n] using t
+end
+
+lemma middling_prime_divides_binom (p : ℕ) (is_prime : prime p) (m : ℕ)
```

Comment:
This name should be something like `dvd_choose_of_middling_prime`, and maybe even `dvd_choose_of_mem_Ico` if you want to use intervals...